lib/puppet/type/file/content.rb

```diff
@@ -119,6 +120,11 @@ def insync?(is)
       result
     end
 
+    def property_matches?(current, desired)
+      return super unless [:mtime, :ctime].include?(checksum_type)
+      DateTime.parse(current) >= DateTime.parse(desired)
```

Comment:
I don't think the super would ever return true however. If either of them doesn't have a time-based checksum, but checksum_type is mtime/ctime, then they're out of sync. That's the behavior you'll get right now.